PeakAnnotationsLoader 1 - Necessary changes to TableLoader base class

hepcat72 opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

I'm working from the tip of a fully working `PeakAnnotationsLoader` class.  Changes in that branch spanned a lot of files, making changes in necessary supporting files, but the change is large, so I am following the following process to break it up:

```
git checkout main
git branch peak_annot_table_loader_1_table_loader
git merge peak_annot_table_loader_rebased  # working branch
git cherry-pick -n 41ca3d434b80c10d5538c668f4308a6354ff57e7  # HEAD of peak_annot_table_loader_1_table_loader
git reset fe501ba73745e13d4d07cd66a2da33b7078e1460  # HEAD of main
git add -p
git commit  # Need to specify new/moved files manually
```

This first PR represents cherry-picked changes related to the `TableLoader` class.  The class file moved (as did `table_column.py`) and a few key changes were made to support the new `PeakAnnotationsLoader`:

Overview:

- Added the ability to take "extra" headers without error (i.e. the sample headers)
- Added a "warned" stat.
- Added the ability to summarize errors based on a SummarizableError class so that the TableLoader code could dynamically summarize any error type without having to know the particulars about an error.

Details:

- TableLoader
  - Location move (wherever it's imported)
    - move TableColumn as well
    - Update wherever either is imported (e.g. other loaders and tests)
  - Exceptions changes
    - SummarizableError is a superclass that effectively simplifies a chunk of `TableLoader` code.  It affects the following exception classes:
      - ConflictingValueErrors
      - DuplicateValueErrors
      - RequiredColumnValues
      - UnknownHeaderError
    - NotATableLoader - This was added to supplant a generic `TypeError` that I found confusing, due to the way the eventual new peak annotations loader class uses new arguments to the constructor.  Table loader will not (as a result) complain if you supply positional arguments (because it takes none and it wasn't complaining when you supplied the wrong arguments positionally).
    - AggregatedErrors - I added a convenience method: `get_exception_types` to make the eventual new code more succinct.
  - A new "warned" stat was added to highlight when there are warnings, because I omitted the `--new-researcher` flag requirement by demoting it to a warning, but I didn't want curators to miss it in the hubbub.  That caused the following loaders/files to have to be updated as well:
    - `DataRepo/loaders/base/table_loader.py`
    - `DataRepo/loaders/lcprotocols_loader.py`
    - `DataRepo/loaders/msruns_loader.py`
    - `DataRepo/loaders/sequences_loader.py`
    - `DataRepo/loaders/tracers_loader.py`
    - `DataRepo/management/commands/load_table.py`

## Affected Issues/Pull Requests

- Resolves #__issue_number__
<!--
- Partially addresses #__issue_number__
- Merges into PR #__pull_request_number__
- Requires merge of PR #__pull_request_number__
-->

## Review Notes
<!--
For added context for reviewers, add comments to relevant lines of code.  Use
this space to describe any general areas of concern not linked to a specific
line of code that reviewers should pay particular attention to, e.g. please
make sure I have accounted for all possible inputs.
-->
See comments in-line.

## Checklist
<!--
If any of the checkbox requirements are not met, uncheck them and add an
explanation. E.g. Linting errors pre-date this PR.
-->
This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 <!-- Edit as desired (e.g. based on complexity) -->
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    <!--
    Require reviewers with specific expertise to be included among the minimum
    number of reviewers by tagging them.  Also please send them a message.
    -->
  - Review period: 2 days <!-- Edit as desired (e.g. based on complexity) -->
- Associated issue/pull request requirements:
  <!--
  Assert that all requirements in issues marked "resolved" are done and that
  all required pull requests are merged.  If any are not done, either edit or
  split the issue or explain the unmerged affected pull requests.
  -->
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  <!--
  Uncheck items to acknowledge failures/conflicts you intend to address.
  Add an explanation if any won't be addressed before merge.
  -->
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  <!--
  These are additional requirements that are not directly associated with
  resolving the affected issue(s).
  -->
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

hepcat72 commented on DataRepo/loaders/__init__.py:
```diff
@@ -3,6 +3,7 @@
     IsotopeObservationData,
```
C. Many of the file changes are due to the location change of the `TableLoader` base class file.  Eventually, there will be a third file added to the new directory.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -12,21 +13,19 @@
-    ConflictingValueErrors,
     DryRun,
     DuplicateHeaders,
-    DuplicateValueErrors,
     DuplicateValues,
     ExcelSheetsNotFound,
     InfileDatabaseError,
     InfileError,
     InvalidHeaderCrossReferenceError,
     NoLoadData,
     RequiredColumnValue,
-    RequiredColumnValues,
     RequiredHeadersError,
     RequiredValueError,
-    RequiredValueErrors,
+    SummarizableError,
+    UnknownHeaderError,
```
C. The old exceptions were retained, but renamed.  We don't need to import these exceptions explicitly anymore because of their association with the new `SummarizableError` superclass.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -177,6 +176,7 @@ def load_data(self):
 
     def __init__(
         self,
+        *args,
```
C. This was added simply to prevent positional arguments, due to confusing behavior when you supply `kwargs` positionally, which becomes an issue later.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -188,6 +188,8 @@ def __init__(
         user_headers=None,
         headers=None,
         defaults=None,
+        extra_headers=None,
```
C. Extra headers were added to allow the sample headers, which are dynamically named.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -188,6 +188,8 @@ def __init__(
         headers=None,
         defaults=None,
+        extra_headers=None,
+        _validate=False,
```
C. This was coming eventually.  This is the same as the historical `validate` argument, which was not needed until recently.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -1971,41 +2014,15 @@ def load_wrapper(self, *args, **kwargs):
-                    if len(cves) > 0:
-                        self.aggregated_errors_object.buffer_error(
-                            ConflictingValueErrors(cves)
-                        )
-
-                    # Summarize any RequiredValueError errors reported
-                    rves = self.aggregated_errors_object.remove_exception_type(
-                        RequiredValueError
-                    )
-                    if len(rves) > 0:
-                        self.aggregated_errors_object.buffer_error(
-                            RequiredValueErrors(rves)
-                        )
-
-                    # Summarize any DuplicateValues errors reported
-                    dvs = self.aggregated_errors_object.remove_exception_type(
-                        DuplicateValues
-                    )
-                    if len(dvs) > 0:
-                        self.aggregated_errors_object.buffer_error(
-                            DuplicateValueErrors(dvs)
-                        )
-
-                    # Summarize any RequiredColumnValue errors reported
-                    rcvs = self.aggregated_errors_object.remove_exception_type(
-                        RequiredColumnValue
-                    )
-                    if len(rcvs) > 0:
-                        self.aggregated_errors_object.buffer_error(
-                            RequiredColumnValues(rcvs)
-                        )
+                    # Summarize multiple types of exceptions that are subclasses of SummarizableError
+                    for exc_cls in self.aggregated_errors_object.get_exception_types():
+                        if issubclass(exc_cls, SummarizableError):
+                            excs = self.aggregated_errors_object.remove_exception_type(
+                                exc_cls
+                            )
+                            self.aggregated_errors_object.buffer_error(
+                                exc_cls.SummarizerExceptionClass(excs)
+                            )
```
C. Not only is this block simplified, but can now handle any exception type that is associated with a summarizable error.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -2142,6 +2159,22 @@ def errored(self, model_name: Optional[str] = None, num=1):
         """
         self.record_counts[self._get_model_name(model_name)]["errored"] += num
 
+    def warned(self, model_name: Optional[str] = None, num=1):
+        """Increments a warned record count for a model.
+
+        Note, this is not for all warnings.  It only pertains to data-specific warnings from the input file.
+
+        Args:
+            model_name (Optional[str])
+
+        Raises:
+            Nothing
+
+        Returns:
+            Nothing
+        """
+        self.record_counts[self._get_model_name(model_name)]["warned"] += num
```
C. The new `warned` stat to mitigate missing issues about researcher variants.  These warnings will be more prominent to users using the web interface.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -2319,32 +2354,96 @@ def handle_load_db_errors(
                     return True
 
         elif isinstance(exception, ValidationError):
```
C. Below, I improved the handling of `ValidationError`s.  Django `ValidationError`s are aggregated exception holders and can contain any number of nested errors.  I was previously handling them all as a single exception without custom exception names (due to Django's repackaging).  This was an issue with the new clean method in `PeakGroup`.  Now, it flattens the nested exceptions (see the `flatten` method at the bottom of this file), and handles them individually.

hepcat72 commented on DataRepo/loaders/tracers_loader.py:
```diff
@@ -207,14 +207,24 @@ def __init__(self, *args, **kwargs):
-                headers (Optional[Tableheaders namedtuple]) [DataHeaders]: Header names by header key.
-                defaults (Optional[Tableheaders namedtuple]) [DataDefaultValues]: Default values by header key.
+                df (Optional[pandas dataframe]): Data, e.g. as parsed from a table-like file.
                 dry_run (Optional[boolean]) [False]: Dry run mode.
                 defer_rollback (Optional[boolean]) [False]: Defer rollback mode.  DO NOT USE MANUALLY - A PARENT SCRIPT
                     MUST HANDLE THE ROLLBACK.
-                sheet (Optional[str]) [None]: Sheet name (for error reporting).
+                data_sheet (Optional[str]) [None]: Sheet name (for error reporting).
+                defaults_sheet (Optional[str]) [None]: Sheet name (for error reporting).
                 file (Optional[str]) [None]: File name (for error reporting).
+                user_headers (Optional[dict]): Header names by header key.
+                defaults_df (Optional[pandas dataframe]): Default values data from a table-like file.
+                defaults_file (Optional[str]) [None]: Defaults file name (None if the same as infile).
+                headers (Optional[DefaultsTableHeaders namedtuple]): headers by header key.
+                defaults (Optional[DefaultsTableHeaders namedtuple]): default values by header key.
+                extra_headers (Optional[List[str]]): Use for dynamic headers (different in every file).  To allow any
+                    unknown header, supply an empty list.
+                _validate (bool): If true, runs in validate mode, perhaps better described as "non-curator mode".  This
+                    is intended for use by the web validation interface.  It's similar to dry-run mode, in that it never
+                    commits anything, but it also raises warnings as fatal (so they can be reported through the web
+                    interface and seen by researchers, among other behaviors specific to non-privileged users).
```
C. This docstring was stale.

hepcat72 commented on DataRepo/management/commands/load_table.py:
```diff
@@ -66,9 +69,12 @@ def __init__(
         *args,
         required_optnames: Optional[List[str]] = None,
         required_optname_groups: Optional[List[List[str]]] = None,
+        opt_defaults: Optional[Dict[str, object]] = None,
```
C. The peak annotations loader needed to set 1 alternate builtin option default.  This allows any derived class to set its own defaults.

hepcat72 commented on DataRepo/management/commands/load_table.py:
```diff
@@ -66,9 +69,12 @@ def __init__(
         required_optnames: Optional[List[str]] = None,
         required_optname_groups: Optional[List[List[str]]] = None,
+        opt_defaults: Optional[Dict[str, object]] = None,
+        custom_loader_init: bool = False,
```
C. The derived classes of peak annotations loader needed to be able to supply `dtype`s to pandas' read methods, and the builtin `dtype` functionality automatically operated on the headers defined in the class, but the class defaults to the "converted" universal format's headers and types, so the initial initialization done before the call to `load_data` needed to be able to forgo the `dtype` option.  By setting a `custom_loader_init` boolean to `True`, it skips the `dtype` functionality (because all it needs is to get the sheet and header names).

hepcat72 commented on DataRepo/management/commands/load_table.py:
```diff
@@ -296,25 +328,17 @@ def check_class_attributes(self):
 
         Args:
             None
-
-        Raises:
-            AggregatedErrors
+        Exceptions:
+            Buffered:
                 TypeError
-
+            Raised:
+                AggregatedErrors
         Returns:
-            Nothing
+            None
         """
-        here = f"{type(self).__module__}.{type(self).__name__}"
         if not issubclass(self.loader_class, TableLoader):
             # Immediately raise programming related errors
-            aes = AggregatedErrors()
-            aes.buffer_error(
-                TypeError(
-                    f"Invalid attribute [{here}.loader_class] TableLoader required, {type(self.loader_class)} set"
-                )
-            )
-            if aes.should_raise():
-                raise aes
+            raise AggregatedErrors().buffer_error(NotATableLoader(self))
```
C. This code is simplified by having the new `NotATableLoader` exception class.

hepcat72 commented on DataRepo/management/commands/load_table.py:
```diff
@@ -376,12 +399,14 @@ def handle_wrapper(self, *args, **options):
 
             # So that derived classes don't need to call init_loader unless their derived loader class takes custom
             # arguments
-            self.init_loader()
+            if not self.custom_loader_init:
+                self.init_loader()
 
             try:
                 retval = fn(self, *args, **options)
 
-            except DryRun:
+            except DryRun as dr:
+                self.dry_run_exception = dr
                 pass
```
C. Before this PR, I had wanted to print a nice dry run message, but raising the dry run exception prevented that.  Now, I actually raise it **and** print my (improved) message, but I don't raise here.  I raise down below after the console print.

hepcat72 commented on DataRepo/management/commands/load_table.py:
```diff
@@ -448,19 +472,29 @@ def report_status(self):
+                # The MIGRATE_HEADING IS BOLD BLUE
+                status = self.style.MIGRATE_HEADING(msg)
+            else:
+                raise CommandError("DryRun exception not raised in --dry-run mode!")
```
C. I added a safe-guard here, just in case.

hepcat72 commented on DataRepo/tests/loaders/test_table_column.py:
C. These 2 test files just moved.

hepcat72 commented on DataRepo/tests/management/test_load_compounds.py:
```diff
@@ -15,6 +15,7 @@
     SynonymExistsAsMismatchedCompound,
```
C. Most of the remaining test changes have to do with either the new DryRun raise or the "no positional args" requirement.

hepcat72 commented on DataRepo/tests/management/test_load_compounds.py:
```diff
@@ -383,18 +384,12 @@ def test_check_for_cross_column_name_duplicates(self):
 @override_settings(CACHES=settings.TEST_CACHES)
 @tag("compound_loading")
 class CompoundValidationLoadingTests(TracebaseTestCase):
-    @classmethod
-    def setUpTestData(cls):
-        call_command(
-            "load_compounds",
-            infile="DataRepo/data/tests/compounds/consolidated_tracebase_compound_list.tsv",
-            dry_run=True,
-            verbosity=0,
-        )
-        # validate only; nothing gets loaded
-        cls.ALL_COMPOUNDS_COUNT = 0
-
-        super().setUpTestData()
```
C. Honestly, I just forgot to put the dry run catch here, but it's tested in multiple other equivalent cases, so this is not really necessary anyway.

hepcat72 commented on DataRepo/tests/management/test_load_table.py:
```diff
@@ -157,16 +159,13 @@ class MyCommand(LoadTableCommand):
+        the wrapper name.
         """
         mc = TestCommand()
-        self.assertFalse(hasattr(mc, "saved_aes"))
-        mc.handle(**self.TEST_OPTIONS)
-        self.assertTrue(hasattr(mc, "saved_aes"))
```
C. I improved the way load_table checks that the handle wrapper has been applied (because I improved the aggregated errors exception handling).  Now it just checks (on the line below) that the method name is literally `handle_wrapper`, (a more direct check instead of inferring based on attributes).

hepcat72 commented on DataRepo/tests/test_models.py:
```diff
@@ -285,19 +285,6 @@ def test_peak_group_unique_constraint(self):
         )
 
 
-@override_settings(CACHES=settings.TEST_CACHES)
-@tag("protocol")
-class ProtocolTests(TracebaseTestCase):
-    def test_create_protocol_by_invalid_category(self):
-        test_protocol_name = "Protocol 2"
-        with self.assertRaises(ValidationError):
-            Protocol.retrieve_or_create_protocol(
-                test_protocol_name,
-                "Invalid Category",
-                "Description",
-            )
```
C. Honestly, this hunk should have been in a different PR.  You'll see that I created a more traditional `get_or_create` for consistency with the other `get_or_creates` I created for the models that the peak annotations loader uses.

hepcat72 commented on DataRepo/utils/exceptions.py:
```diff
@@ -133,29 +165,6 @@ class HeaderError(Exception):
     pass
 
 
```
C. Moved lower to associate with a summarizable error.

hepcat72 commented on DataRepo/utils/exceptions.py:
```diff
@@ -199,16 +208,29 @@ def __init__(
 
-class RequiredColumnValue(InfileError):
+class RequiredValueError(InfileError, SummarizableError):
+    SummarizerExceptionClass = RequiredValueErrors
```
C. Repeated exceptions with a `SummarizerExceptionClass` attribute are automatically summarized by `TableLoader`.  This was useful when I wanted to extend the design pattern Lance established of having a class that takes a list of exceptions of a certain type that I really liked.  I wanted to do this for repeated exceptions in the new peak annotation loader without adding all thos exceptions to `TableLoader` explicitly, because it shouldn't know about the specific exceptions it raises.

hepcat72 commented on DataRepo/utils/exceptions.py:
```diff
@@ -1650,22 +1741,44 @@ def __init__(self, dupe_dict, colnames, message=None, addendum=None, **kwargs):
         self.dupdeets = dupdeets
 
 
```
C. `DuplicateCompoundIsotopes` probably should have been in a different hunk, but when I did the first pass, I'd thought this was one of the summarizable exceptions since it was in the same place as `DuplicateValueErrors`, which is summarizable.

lparsons reviewed (approved):
Nice work updating the TableLoader class. I added a few comments in-line, but saw nothing blocking. Merge when you feel it's ready.

One item I thought about when looking this over was the `DryRun` exception. I don't feel that the management command should be raising a `DryRun` exception if the `--dry-run` flag is set. Instead, I think there should be an underlying call that raises the exception to roll things back and the command should exit successfully (without raising an exception). This isn't something I feel needs to change in this PR, but it's a general strategy I'd consider for all management commands.

hepcat72 commented on DataRepo/loaders/accucor_data_loader.py:
```diff
@@ -4,14 +4,14 @@
 from django.core.exceptions import ObjectDoesNotExist, ValidationError
 from django.core.files import File
 from django.db import IntegrityError, transaction
+from django.db.utils import ProgrammingError
```
Yeah, I suspect that the sqlite3 one got in there because of VS Code's auto-import feature. I don't know if you use that, but as you type a method or class, if it's not among the currently imported stuff, you get a dropdown of different packages and if you hit enter, it automatically adds an import for the first one.  I've gotten more used to the feature now, and don't accidentally add imports as much.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -206,13 +208,26 @@ def __init__(
         Returns:
             Nothing
         """
+        if len(args) > 0:
+            raise AggregatedErrors().buffer_error(
+                ProgrammingError(
+                    f"The TableLoader constructor expects 0 positional arguments, but got: {len(args)}."
+                ),
+            )
+
```
Oh, nifty.

- [ ] TODO: Create an issue to employ the use of `*` when you require 0 positional args.

hepcat72 commented on DataRepo/loaders/base/table_loader.py:
```diff
@@ -2319,32 +2354,96 @@ def handle_load_db_errors(
+                            # ignoring a duplicate)
+                            return True
+
+                elif issubclass(
+                    type(orig_exception), ValidationError
+                ) and not issubclass(type(orig_exception), InfileError):
```
That's qualitatively different from what is required here.  While your suggestion will indeed match subclasses of `ValidationError`, it will also match direct instances of `ValidationError`, but the code in use here specifically does not match direct instances of `ValidationError`.  It's intended to only match subclasses of `ValidationError` (that are also not subclasses of `InfileError`).

hepcat72 commented:
> One item I thought about when looking this over was the `DryRun` exception. I don't feel that the management command should be raising a `DryRun` exception if the `--dry-run` flag is set. Instead, I think there should be an underlying call that raises the exception to roll things back and the command should exit successfully (without raising an exception). This isn't something I feel needs to change in this PR, but it's a general strategy I'd consider for all management commands.

Yeah.  I got confused when I was testing, because I think I'd gotten used to seeing dry run exceptions on the command line, and the old accucor data loader was raising DryRun exceptions, so when I copied and updated accucor loader tests, my tests failed because I wasn't raising the dry run exception.  It also was confusing because the status messages were green and I was like: was the dry run exception raised or not?

In general, I'm still not entirely satisfied with the way Python handles exceptions.  I think they're quite powerful and open up a ton of possibilities, but I hate stack traces.  They're ugly, I find their presentation order to be backwards, their "during" messages aren't always correct (i.e. the description preceding the sub-trace incorrectly describes the order of operations sometimes), and you can't get (a true) one if you continue to allow the script to run - but they're just so darned useful for debugging.  I really liked the more succinct custom traces I implemented in my perl work.

I guess I just got used to python exceptions, despite the things I dislike about them.  But I agree with you.  I think that it shouldn't be raised from the command script, and I think that the color change I implemented for dry run messages should suffice to clearly convey what happened.

- [ ] TODO: Create an issue to NOT raise `DryRun` exceptions from the command scripts.
